### DISCUSSION TITLE
Debugging gem5: BrokenPipeError

### DISCUSSION THREAD

**Discussion by zhaorui233** [gem5/gem5#227] [Q&A] [answered]
I follow the book chapter 'Debugging gem5'
run cmd : build/X86/gem5.opt --debug-flags=Exec configs/learning_gem5/part1/simple.py | head -n 50
shows:
Global frequency set at 1000000000000 ticks per second
warn: No dot file generated. Please install pydot to generate the dot file and pdf.
src/mem/dram_interface.cc:690: warn: DRAM device capacity (8192 Mbytes) does not match the address range assigned (512 Mbytes)
src/base/statistics.hh:279: warn: One of the stats is a legacy stat. Legacy stat is a stat that does not belong to any statistics::Group. Legacy stat is deprecated.
system.remote_gdb: Listening for connections on port 7000
src/sim/simulate.cc:194: info: Entering event queue @ 0.  Starting simulation...
gem5 Simulator System.  https://www.gem5.org
gem5 is copyrighted software; use the --copyright option for details.

gem5 version 23.0.1.0
gem5 compiled Aug 24 2023 15:14:23
gem5 started Aug 24 2023 15:18:10
gem5 executing on dell-workstation, pid 68707
command line: build/X86/gem5.opt --debug-flags=Exec configs/learning_gem5/part1/simple.py

Beginning simulation!
  77000: system.cpu: T0 : 0x400190 @_start    : xor	rbp                   
  77000: system.cpu: T0 : 0x400190 @_start. 0 :   XOR_R_R : xor   rbp, rbp, rbp : IntAlu :  D=0x0000000000000000
 126000: system.cpu: T0 : 0x400193 @_start+3    : mov	rdx                   
 126000: system.cpu: T0 : 0x400193 @_start+3. 0 :   MOV_R_R : mov   r9, r9, rdx : IntAlu :  D=0x0000000000000000
 175000: system.cpu: T0 : 0x400196 @_start+6    : pop	rsi                   
 175000: system.cpu: T0 : 0x400196 @_start+6. 0 :   POP_R : ld   t1, SS:[t0 + rsp] : MemRead :  D=0x0000000000000001 A=0x7fffffffed90
 238000: system.cpu: T0 : 0x400196 @_start+6. 1 :   POP_R : addi   rsp, rsp, 0x8 : IntAlu :  D=0x00007fffffffed98
 238000: system.cpu: T0 : 0x400196 @_start+6. 2 :   POP_R : mov   rsi, rsi, t1 : IntAlu :  D=0x0000000000000001
 336000: system.cpu: T0 : 0x400197 @_start+7    : mov	rsp                   
 336000: system.cpu: T0 : 0x400197 @_start+7. 0 :   MOV_R_R : mov   rdx, rdx, rsp : IntAlu :  D=0x00007fffffffed98
 385000: system.cpu: T0 : 0x40019a @_start+10    : and	rsp, 0xfffffffffffffff0
 385000: system.cpu: T0 : 0x40019a @_start+10. 0 :   AND_R_I : limm   t1, 0xfffffffffffffff0 : IntAlu :  D=0xfffffffffffffff0
 385000: system.cpu: T0 : 0x40019a @_start+10. 1 :   AND_R_I : and   rsp, rsp, t1 : IntAlu :  D=0x0000000000000000
 434000: system.cpu: T0 : 0x40019e @_start+14    : push	rax                  
 434000: system.cpu: T0 : 0x40019e @_start+14. 0 :   PUSH_R : st   rax, SS:[t0 + rsp + 0xfffffffffffffff8] : MemWrite :  D=0x0000000000000000 A=0x7fffffffed88
 455000: system.cpu: T0 : 0x40019e @_start+14. 1 :   PUSH_R : subi   rsp, rsp, 0x8 : IntAlu :  D=0x00007fffffffed88
 504000: system.cpu: T0 : 0x40019f @_start+15    : push	rsp                  
 504000: system.cpu: T0 : 0x40019f @_start+15. 0 :   PUSH_R : st   rsp, SS:[t0 + rsp + 0xfffffffffffffff8] : MemWrite :  D=0x00007fffffffed88 A=0x7fffffffed80
 525000: system.cpu: T0 : 0x40019f @_start+15. 1 :   PUSH_R : subi   rsp, rsp, 0x8 : IntAlu :  D=0x00007fffffffed80
 574000: system.cpu: T0 : 0x4001a0 @_start+16    : mov	r15, 0x40a060         
 574000: system.cpu: T0 : 0x4001a0 @_start+16. 0 :   MOV_R_I : limm   r8, 0x40a060 : IntAlu :  D=0x000000000040a060
 672000: system.cpu: T0 : 0x4001a7 @_start+23    : mov	rdi, 0x409ff0         
 672000: system.cpu: T0 : 0x4001a7 @_start+23. 0 :   MOV_R_I : limm   rcx, 0x409ff0 : IntAlu :  D=0x0000000000409ff0
 770000: system.cpu: T0 : 0x4001ae @_start+30    : mov	rdi, 0x400274         
 770000: system.cpu: T0 : 0x4001ae @_start+30. 0 :   MOV_R_I : limm   rdi, 0x400274 : IntAlu :  D=0x0000000000400274
 868000: system.cpu: T0 : 0x4001b5 @_start+37    : call	0x9846               
 868000: system.cpu: T0 : 0x4001b5 @_start+37. 0 :   CALL_NEAR_I : limm   t1, 0x9846 : IntAlu :  D=0x0000000000009846
 868000: system.cpu: T0 : 0x4001b5 @_start+37. 1 :   CALL_NEAR_I : rdip   t7, t7 : IntAlu :  D=0x00000000004001ba
 868000: system.cpu: T0 : 0x4001b5 @_start+37. 2 :   CALL_NEAR_I : st   t7, SS:[rsp + 0xfffffffffffffff8] : MemWrite :  D=0x00000000004001ba A=0x7fffffffed78
 889000: system.cpu: T0 : 0x4001b5 @_start+37. 3 :   CALL_NEAR_I : subi   rsp, rsp, 0x8 : IntAlu :  D=0x00007fffffffed78
 889000: system.cpu: T0 : 0x4001b5 @_start+37. 4 :   CALL_NEAR_I : wrip   t7, t1 : IntAlu : 
 952000: system.cpu: T0 : 0x409a00 @__libc_start_main    : push	r15                  
 952000: system.cpu: T0 : 0x409a00 @__libc_start_main. 0 :   PUSH_R : st   r15, SS:[t0 + rsp + 0xfffffffffffffff8] : MemWrite :  D=0x0000000000000000 A=0x7fffffffed70
 973000: system.cpu: T0 : 0x409a00 @__libc_start_main. 1 :   PUSH_R : subi   rsp, rsp, 0x8 : IntAlu :  D=0x00007fffffffed70
1022000: system.cpu: T0 : 0x409a02 @__libc_start_main+2    : movsxd	rax                
1022000: system.cpu: T0 : 0x409a02 @__libc_start_main+2. 0 :   MOVSXD_R_R : sexti   rax, rsi, 0x1f : IntAlu :  D=0x0000000000000001
1071000: system.cpu: T0 : 0x409a05 @__libc_start_main+5    : mov	r9                    
1071000: system.cpu: T0 : 0x409a05 @__libc_start_main+5. 0 :   MOV_R_R : mov   r15, r15, r9 : IntAlu :  D=0x0000000000000000
1120000: system.cpu: T0 : 0x409a08 @__libc_start_main+8    : push	r14                  
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='ascii'>
BrokenPipeError: [Errno 32] Broken pipe

how to solve this error ?

**Comment by abmerop** [comment, marked as answer]
Remove the pipe to `head -n 50`.  The error is Linux telling you the pipe'd process (gem5) is exiting because the program receiving the output ended and also why there are exactly 50 lines of output

**Reply by zhaorui233** [reply]
How to remove the pipe ? Should I use some commands or write a script ?

**Reply by BobbyRBruce** [reply]
The command you ran was `build/X86/gem5.opt --debug-flags=Exec configs/learning_gem5/part1/simple.py  | head -n 50`.  The  `|` is the "pipe". It sends the output of `build/X86/gem5.opt --debug-flags=Exec configs/learning_gem5/part1/simple.py` to the `head` application with the arguments `-n 50`. `head -n 50` will output the first 50 lines it gets then stop. When this stop occurs gem5 crashes as it's got nowhere left to output to.

So, all you need to do is execute `build/X86/gem5.opt --debug-flags=Exec configs/learning_gem5/part1/simple.py` (i.e., without the trailing `| head -n 50`).

**Reply by zhaorui233** [reply]
Thanks for replying. It works now.
